src/operator/distinct.rs

```diff
@@ -125,86 +160,402 @@ where
+        self.empty_input
+            && self.empty_output
+            && self
+                .future_updates
+                .iter()
+                .skip(self.time as usize)
+                .all(|vals| vals.is_empty())
     }
 
-    fn eval_owned(&mut self, delta: Z, delayed_integral: Z) -> Z {
-        self.eval_owned_and_ref(delta, &delayed_integral)
+    fn summary(&self, summary: &mut String) {
+        let size: usize = self.future_updates.iter().map(|vals| vals.len()).sum();
+        writeln!(summary, "size: {}", size).unwrap();
+
+        let bytes = self.future_updates.deep_size_of();
+        writeln!(summary, "bytes: {}", bytes).unwrap();
+    }
+}
+
+impl<Z, T> BinaryOperator<Z, T, Z> for DistinctTrace<Z, T>
+where
+    Z: ZSet,
+    Z::Key: Clone + Ord + PartialEq + DeepSizeOf,
+    Z::R: ZRingValue,
+    T: Trace<Key = Z::Key, Val = (), Time = NestedTimestamp32, R = Z::R> + 'static,
+{
+    // Distinct does not have nice properties like linearity that help with
+    // incremental evaluation: computing an update for each key requires inspecting
+    // the entire history of this key.  Fortunately, we do not need to look at all
+    // keys in the trace.  In particular, only those keys that appeared in input
+    // deltas during the current clock epoch can appear in the output of this
+    // operator with non-zero weights.  So, as a simple heuristic, we could track
+    // the set of keys that appeared since the start of the epoch and only consider
+    // those.
+    //
+    // But we can restrict the set of keys to consider even more by using a stronger
+    // property: only the keys `k` that satisfy one of the following conditions can
+    // appear in the output of the operator at local time `t`:
+    // * keys in the current input `delta`
+    // * keys from earlier inputs observed during the current clock epoch that
+    //   appeared in one of the previous epochs at time `t`.
+    //
+    // To efficiently compute keys that satisfy the second condition, we use the
+    // `future_updates` vector, where for each key observed in the current epoch
+    // at time `t1` we lookup the smallest time `t2 > t1` (if any) at which we saw
+    // the key during any previous epochs and record this key in
+    // `future_updates[t2]`. Then when evaluating an operatpr at time `t` we
+    // simply scan all keys in `delta` and all keys in `future_updates[t]` for
+    // candidates.
+    //
+    // TODO: Add example.
+    //
+    fn eval(&mut self, delta: &Z, trace: &T) -> Z {
+        //eprintln!("distinct_trace {}", self.time);
+
+        self.empty_input = delta.is_zero();
+
+        // Make sure we have enough room in `future_updates` to
+        // accommodate the largest timestamp in the trace, so we don't
+        // need to worry about growing `future_updates` later on.
+        let mut new_len: u32 = self.time + 1;
+        trace.map_batches(|batch| {
+            for ts in batch.upper().elements().iter() {
```

Comment:
It's not really incremental.  I just scan all the batches searching for the largest timestamp and make sure I have enough room in the array.  Since we generally don't have an upper bound of the number of iterations, we can't allocate a big enough array at initialization time.